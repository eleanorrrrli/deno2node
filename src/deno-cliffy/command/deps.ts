export {
  bold,
  brightBlue,
  brightMagenta,
  cyan,
  dim,
  getColorEnabled,
  green,
  italic,
  red,
  setColorEnabled,
  yellow,
} from "https://deno.land/std@0.221.0/fmt/colors.ts";
