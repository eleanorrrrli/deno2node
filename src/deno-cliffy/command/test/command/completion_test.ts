// deno-fmt-ignore-file

import { assertEquals } from "../../../dev_deps.ts";
import { snapshotTest } from "../../../testing/snapshot.ts";
import { Command } from "../../command.ts";
import { CompletionsCommand } from "../../completions/completions_command.ts";
import type { Completion } from "../../types.ts";

function command() {
  return new Command()
    .throwErrors()
    .globalComplete("global", () => ["global1", "global2"])
    .complete("foo", () => ["foo1", "foo2"])
    .command(
      "bar",
      new Command()
        .complete("bar", () => ["bar1", "bar2"])
        .command("baz")
        .complete("baz", () => ["baz1", "baz2"]),
    );
}

Deno.test("command - completion - completion properties", () => {
  const cmd = new Command()
    .throwErrors()
    .complete("foo", () => ["foo1", "foo2"])
    .complete("bar", () => ["bar1", "bar2"], {
      global: true,
      override: true,
    });
  const fooCompletion: Completion = cmd.getCompletion("foo") as Completion;
  const barCompletion: Completion = cmd.getCompletion("bar") as Completion;
  assertEquals(fooCompletion.name, "foo");
  assertEquals(fooCompletion.global, undefined);
  assertEquals(barCompletion.name, "bar");
  assertEquals(barCompletion.global, true);
});

Deno.test("command - completion - get completions", () => {
  const cmd = command();
  assertEquals(cmd.getCompletions().length, 2);
  assertEquals(!!cmd.getCompletions().find((opt) => opt.name === "global"), true);
  assertEquals(!!cmd.getCompletions().find((opt) => opt.name === "foo"), true);
});

Deno.test("command - completion - get base completions", () => {
  const cmd = command();
  assertEquals(cmd.getBaseCompletions().length, 2);
  assertEquals(!!cmd.getBaseCompletions().find((opt) => opt.name === "global"), true);
  assertEquals(!!cmd.getBaseCompletions().find((opt) => opt.name === "foo"), true);
});

Deno.test("command - completion - get global completions", () => {
  const cmd = command();
  assertEquals(cmd.getCommand("bar")?.getGlobalCompletions().length, 1);
  assertEquals(!!cmd.getCommand("bar")?.getGlobalCompletions().find((opt) => opt.name === "global"), true);
  assertEquals(!!cmd.getCommand("bar")?.getGlobalCompletions().find((opt) => opt.name === "foo"), false);
});

Deno.test("command - completion - get completion", () => {
  const cmd = command();
  assertEquals(cmd.getCompletion("global")?.name, "global");
  assertEquals(cmd.getCompletion("foo")?.name, "foo");
  assertEquals(cmd.getCompletion("unknown")?.name, undefined);
  assertEquals(cmd.getCommand("bar")?.getCompletion("global")?.name, "global");
  assertEquals(cmd.getCommand("bar")?.getCompletion("bar")?.name, "bar");
  assertEquals(cmd.getCommand("bar")?.getCompletion("unknown")?.name, undefined);
  assertEquals(cmd.getCommand("bar")?.getCommand("baz")?.getCompletion("global")?.name, "global");
  assertEquals(cmd.getCommand("bar")?.getCommand("baz")?.getCompletion("baz")?.name, "baz");
  assertEquals(cmd.getCommand("bar")?.getCommand("baz")?.getCompletion("unknown")?.name, undefined);
});

Deno.test("command - completion - get base completion", () => {
  const cmd = command();
  assertEquals(cmd.getBaseCompletion("global")?.name, "global");
  assertEquals(cmd.getBaseCompletion("foo")?.name, "foo");
  assertEquals(cmd.getBaseCompletion("unknown")?.name, undefined);
  assertEquals(cmd.getCommand("bar")?.getBaseCompletion("global")?.name, undefined);
  assertEquals(cmd.getCommand("bar")?.getBaseCompletion("bar")?.name, "bar");
  assertEquals(cmd.getCommand("bar")?.getBaseCompletion("unknown")?.name, undefined);
  assertEquals(cmd.getCommand("bar")?.getCommand("baz")?.getBaseCompletion("global")?.name, undefined);
  assertEquals(cmd.getCommand("bar")?.getCommand("baz")?.getBaseCompletion("baz")?.name, "baz");
  assertEquals(cmd.getCommand("bar")?.getCommand("baz")?.getBaseCompletion("unknown")?.name, undefined);
});

Deno.test("command - completion - get global completion", () => {
  const cmd = command();
  assertEquals(cmd.getGlobalCompletion("global")?.name, undefined);
  assertEquals(cmd.getGlobalCompletion("foo")?.name, undefined);
  assertEquals(cmd.getGlobalCompletion("unknown")?.name, undefined);
  assertEquals(cmd.getCommand("bar")?.getGlobalCompletion("global")?.name, "global");
  assertEquals(cmd.getCommand("bar")?.getGlobalCompletion("bar")?.name, undefined);
  assertEquals(cmd.getCommand("bar")?.getGlobalCompletion("unknown")?.name, undefined);
  assertEquals(cmd.getCommand("bar")?.getCommand("baz")?.getGlobalCompletion("global")?.name, "global");
  assertEquals(cmd.getCommand("bar")?.getCommand("baz")?.getGlobalCompletion("baz")?.name, undefined);
  assertEquals(cmd.getCommand("bar")?.getCommand("baz")?.getGlobalCompletion("unknown")?.name, undefined);
});

await snapshotTest({
  name: "should handle file type completions in options",
  meta: import.meta,
  steps: {
    zsh: { args: ["completions", "zsh"] },
  },
  async fn(): Promise<void> {
    await new Command()
      .version("1.0.0")
      .name("completions-test")
      // first option argument
      .command("foo")
      .option("-f, --file1 <path:file>", "...", { required: true })
      // second option argument
      .command("bar")
      .option("-f, --file2 <name:string> <path:file>", "...", { required: true })
      // rest option argument
      .command("beep")
      .option("-f, --file3 <...path:file>", "...", { required: true })
      // second rest option argument
      .command("boop")
      .option("-f, --file4 <name:string> <...path:file>", "...", { required: true })
      // completion command
      .command("completions", new CompletionsCommand())
      .parse();
  },
});

await snapshotTest({
  name: "should generate file type completions in arguments",
  meta: import.meta,
  steps: {
    zsh: { args: ["completions", "zsh"] },
  },
  async fn(): Promise<void> {
    await new Command()
      .version("1.0.0")
      .name("completions-test")
      // first argument
      .command("foo")
      .arguments("<path:file>")
      // second argument
      .command("bar")
      .arguments("<name:string> <path:file>")
      // rest argument
      .command("beep")
      .arguments("<...path:file>")
      // second rest argument
      .command("boop")
      .arguments("<name:string> <...path:file>")
      // completion command
      .command("completions", new CompletionsCommand())
      .parse();
  },
});
