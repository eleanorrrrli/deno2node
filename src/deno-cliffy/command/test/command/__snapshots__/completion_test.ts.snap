export const snapshot = {};

snapshot[`should handle file type completions in options > zsh 1`] = `
stdout:
\`#!/usr/bin/env zsh
# zsh completion support for completions-test v1.0.0

autoload -U is-at-least

# shellcheck disable=SC2154
(( \$+functions[__completions_test_complete] )) ||
function __completions_test_complete {
  local name="\$1"; shift
  local action="\$1"; shift
  integer ret=1
  local -a values
  local expl lines
  _tags "\$name"
  while _tags; do
    if _requested "\$name"; then
      # shellcheck disable=SC2034
      lines="\$(completions-test completions complete "\${action}" "\${@}")"
      values=("\${(ps:\\\\n:)lines}")
      if (( \${#values[@]} )); then
        while _next_label "\$name" expl "\$action"; do
          compadd -S '' "\${expl[@]}" "\${values[@]}"
        done
      fi
    fi
  done
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test] )) ||
function _completions_test() {
  local state

  function _commands() {
    local -a commands
    # shellcheck disable=SC2034
    commands=(
      'foo:'
      'bar:'
      'beep:'
      'boop:'
      'completions:Generate shell completions.'
    )
    _describe 'command' commands
  }

  function _command_args() {
    case "\${words[1]}" in
      foo) _completions_test_foo ;;
      bar) _completions_test_bar ;;
      beep) _completions_test_beep ;;
      boop) _completions_test_boop ;;
      completions) _completions_test_completions ;;
    esac
  }

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(- *)'{-V,--version}'[Show the version number for this program.]' \\\\
    '1:command:_commands' \\\\
    '*::sub command:->command_args'

  case "\$state" in
    command_args) _command_args ;;
  esac
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_foo] )) ||
function _completions_test_foo() {

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(-h --help -f --file1)'{-f,--file1}'[...]:path:_files'
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_bar] )) ||
function _completions_test_bar() {

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(-h --help -f --file2)'{-f,--file2}'[...]:name:->name-string:path:_files'

  case "\$state" in
    name-string) __completions_test_complete name string bar ;;
  esac
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_beep] )) ||
function _completions_test_beep() {

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(-h --help -f --file3)'{-f,--file3}'[...]:path:_files'
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_boop] )) ||
function _completions_test_boop() {

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(-h --help -f --file4)'{-f,--file4}'[...]:name:->name-string:path:_files'

  case "\$state" in
    name-string) __completions_test_complete name string boop ;;
  esac
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_completions] )) ||
function _completions_test_completions() {

  function _commands() {
    local -a commands
    # shellcheck disable=SC2034
    commands=(
      'bash:Generate shell completions for bash.'
      'fish:Generate shell completions for fish.'
      'zsh:Generate shell completions for zsh.'
    )
    _describe 'command' commands
  }

  function _command_args() {
    case "\${words[1]}" in
      bash) _completions_test_completions_bash ;;
      fish) _completions_test_completions_fish ;;
      zsh) _completions_test_completions_zsh ;;
    esac
  }

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '1:command:_commands' \\\\
    '*::sub command:->command_args'

  case "\$state" in
    command_args) _command_args ;;
  esac
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_completions_bash] )) ||
function _completions_test_completions_bash() {

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(-h --help -n --name)'{-n,--name}'[The name of the main command.]:command-name:->command-name-string'

  case "\$state" in
    command-name-string) __completions_test_complete command-name string completions bash ;;
  esac
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_completions_fish] )) ||
function _completions_test_completions_fish() {

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(-h --help -n --name)'{-n,--name}'[The name of the main command.]:command-name:->command-name-string'

  case "\$state" in
    command-name-string) __completions_test_complete command-name string completions fish ;;
  esac
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_completions_zsh] )) ||
function _completions_test_completions_zsh() {

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(-h --help -n --name)'{-n,--name}'[The name of the main command.]:command-name:->command-name-string'

  case "\$state" in
    command-name-string) __completions_test_complete command-name string completions zsh ;;
  esac
}

# _completions_test "\${@}"

compdef _completions_test completions-test
\`
stderr:
""
`;

snapshot[`should generate file type completions in arguments > zsh 1`] = `
stdout:
\`#!/usr/bin/env zsh
# zsh completion support for completions-test v1.0.0

autoload -U is-at-least

# shellcheck disable=SC2154
(( \$+functions[__completions_test_complete] )) ||
function __completions_test_complete {
  local name="\$1"; shift
  local action="\$1"; shift
  integer ret=1
  local -a values
  local expl lines
  _tags "\$name"
  while _tags; do
    if _requested "\$name"; then
      # shellcheck disable=SC2034
      lines="\$(completions-test completions complete "\${action}" "\${@}")"
      values=("\${(ps:\\\\n:)lines}")
      if (( \${#values[@]} )); then
        while _next_label "\$name" expl "\$action"; do
          compadd -S '' "\${expl[@]}" "\${values[@]}"
        done
      fi
    fi
  done
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test] )) ||
function _completions_test() {
  local state

  function _commands() {
    local -a commands
    # shellcheck disable=SC2034
    commands=(
      'foo:'
      'bar:'
      'beep:'
      'boop:'
      'completions:Generate shell completions.'
    )
    _describe 'command' commands
  }

  function _command_args() {
    case "\${words[1]}" in
      foo) _completions_test_foo ;;
      bar) _completions_test_bar ;;
      beep) _completions_test_beep ;;
      boop) _completions_test_boop ;;
      completions) _completions_test_completions ;;
    esac
  }

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(- *)'{-V,--version}'[Show the version number for this program.]' \\\\
    '1:command:_commands' \\\\
    '*::sub command:->command_args'

  case "\$state" in
    command_args) _command_args ;;
  esac
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_foo] )) ||
function _completions_test_foo() {

  function _commands() {
    _files
  }

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '1:command:_commands'
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_bar] )) ||
function _completions_test_bar() {

  function _commands() {
  }

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '1:command:_commands'\\\\
    '2:path:_files'
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_beep] )) ||
function _completions_test_beep() {

  function _commands() {
    _files
  }

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '*:command:_commands'
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_boop] )) ||
function _completions_test_boop() {

  function _commands() {
  }

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '1:command:_commands'\\\\
    '2:path:_files'\\\\
    '3:path:_files'\\\\
    '4:path:_files'\\\\
    '5:path:_files'\\\\
    '6:path:_files'
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_completions] )) ||
function _completions_test_completions() {

  function _commands() {
    local -a commands
    # shellcheck disable=SC2034
    commands=(
      'bash:Generate shell completions for bash.'
      'fish:Generate shell completions for fish.'
      'zsh:Generate shell completions for zsh.'
    )
    _describe 'command' commands
  }

  function _command_args() {
    case "\${words[1]}" in
      bash) _completions_test_completions_bash ;;
      fish) _completions_test_completions_fish ;;
      zsh) _completions_test_completions_zsh ;;
    esac
  }

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '1:command:_commands' \\\\
    '*::sub command:->command_args'

  case "\$state" in
    command_args) _command_args ;;
  esac
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_completions_bash] )) ||
function _completions_test_completions_bash() {

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(-h --help -n --name)'{-n,--name}'[The name of the main command.]:command-name:->command-name-string'

  case "\$state" in
    command-name-string) __completions_test_complete command-name string completions bash ;;
  esac
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_completions_fish] )) ||
function _completions_test_completions_fish() {

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(-h --help -n --name)'{-n,--name}'[The name of the main command.]:command-name:->command-name-string'

  case "\$state" in
    command-name-string) __completions_test_complete command-name string completions fish ;;
  esac
}

# shellcheck disable=SC2154
(( \$+functions[_completions_test_completions_zsh] )) ||
function _completions_test_completions_zsh() {

  _arguments -w -s -S -C \\\\
    '(- *)'{-h,--help}'[Show this help.]' \\\\
    '(-h --help -n --name)'{-n,--name}'[The name of the main command.]:command-name:->command-name-string'

  case "\$state" in
    command-name-string) __completions_test_complete command-name string completions zsh ;;
  esac
}

# _completions_test "\${@}"

compdef _completions_test completions-test
\`
stderr:
""
`;
